File: valid_palindrome_ii/ValidPalindromeII.java
class ValidPalindromeII {
    public boolean validPalindrome(String s) {
        if (s == null) {
            return false;
        }
        if (s.length() < 3) {
            return true;
        }
        
        int left = 0;
        int right = s.length() - 1;
        while (left <= right && s.charAt(left++) == s.charAt(right--));
        return left > right || validPalindrome(s, left, right + 1) || validPalindrome(s, left - 1, right);
    }
    
    private boolean validPalindrome(String s, int left, int right) {
        while (left <= right && s.charAt(left++) == s.charAt(right--));
        return left > right && s.charAt(left - 1) == s.charAt(right + 1);
    }
}

File: valid_palindrome_ii/Valid_Palindrome_II.cpp
class Valid_Palindrome_II {
public:
    bool validPalindrome(string s) {
        if (s.empty()) {
            return true;
        }
        
        int left = 0;
        int right = s.size() - 1;
        while (left <= right) {
            if (s[left++] == s[right--]) {
                continue;
            }
            return validPalindrome(s, left, right + 1) || validPalindrome(s, left - 1, right);
        }
        
        return true;
    }
private:
    bool validPalindrome(string s, int left, int right) {
        while (left <= right && s[left] == s[right]) {
            left++;
            right--;
        }
        return left > right;
    }
};